librz/core/cmd/cmd_search_rop.c

```diff
@@ -4,726 +4,402 @@
-	return list;
+static bool parse_eof(const char *str, int idx) {
+	skip_whitespace(str, &idx);
+	return str[idx] == '\0';
 }
 
-static RzList /*<char *>*/ *get_constants(const char *str) {
-	RzList *list;
-	char *p, *data;
-	if (!str) {
-		return NULL;
+static bool parse_il_equal(char *str, int *idx) {
+	skip_whitespace(str, idx);
+	if (*idx >= strlen(str)) {
+		return false;
 	}
-
-	data = strdup(str);
-	list = rz_list_newf(free);
-	p = strtok(data, ",");
-	while (p) {
-		if (strtol(p, NULL, 0)) {
-			rz_list_append(list, (void *)strdup(p));
-		}
-		p = strtok(NULL, ",");
+	if (str[*idx] == '=') {
+		(*idx)++;
+		return true;
 	}
-	free(data);
-	return list;
+	return false;
 }
 
-static bool isFlag(RzRegItem *reg) {
-	const char *type = rz_reg_get_type(reg->type);
+static char *parse_register(RzCore *core, char *str, int *idx) {
+	char reg[256] = { 0 };
+	int reg_idx = 0;
 
-	if (!strcmp(type, "flg"))
-		return true;
-	return false;
-}
+	skip_whitespace(str, idx);
 
-// binary op
-static bool simulate_op(const char *op, ut64 src1, ut64 src2, ut64 old_src1, ut64 old_src2, ut64 *result, int size) {
-	ut64 limit;
-	if (size == 64) {
-		limit = UT64_MAX;
-	} else {
-		limit = 1ULL << size;
+	while (isalnum(str[*idx]) || str[*idx] == '_') {
+		reg[reg_idx++] = str[*idx];
+		(*idx)++;
 	}
 
-	if (!strcmp(op, "^")) {
-		*result = src1 ^ src2;
-		return true;
-	}
-	if (!strcmp(op, "+")) {
-		*result = src1 + src2;
-		return true;
-	}
-	if (!strcmp(op, "-")) {
-		if (src2 > src1) {
-			*result = limit + (src1 - src2);
-		} else {
-			*result = src1 - src2;
-		}
-		return true;
-	}
-	if (!strcmp(op, "*")) {
-		*result = src1 * src2;
-		return true;
-	}
-	if (!strcmp(op, "|")) {
-		*result = src1 | src2;
-		return true;
-	}
-	if (!strcmp(op, "/")) {
-		*result = src1 / src2;
-		return true;
-	}
-	if (!strcmp(op, "%")) {
-		*result = src1 % src2;
-		return true;
-	}
-	if (!strcmp(op, "<<")) {
-		*result = src1 << src2;
-		return true;
-	}
-	if (!strcmp(op, ">>")) {
-		*result = src1 >> src2;
-		return true;
-	}
-	if (!strcmp(op, "&")) {
-		*result = src1 & src2;
-		return true;
+	if (reg_idx == 0) {
+		return NULL;
 	}
-	if (!strcmp(op, "+=")) {
-		*result = old_src1 + src2;
-		return true;
+
+	// Check if the register is correct for the given architecture.
+	if (rz_analysis_is_reg_in_profile(core->analysis, reg)) {
+		return strdup(reg);
 	}
-	if (!strcmp(op, "-=")) {
-		if (src2 > old_src1) {
-			*result = limit + (old_src1 - src2);
-		} else {
-			*result = old_src1 - src2;
-		}
-		return true;
+
+	return NULL;
+}
+
+static bool parse_constant(const char *str, int *idx, unsigned long long *value) {
+	int base = 10;
+	int neg = 0;
+	char num_str[256] = { 0 };
+	int num_idx = 0;
+
+	skip_whitespace(str, idx);
+
+	if (str[*idx] == '-') {
+		neg = 1;
+		(*idx)++;
 	}
-	if (!strcmp(op, "*=")) {
-		*result = old_src1 * src2;
-		return true;
+
+	skip_whitespace(str, idx);
+
+	if (str[*idx] == '0' && (str[*idx + 1] == 'x' || str[*idx + 1] == 'X')) {
```

Comment:
this might be an out of bound read. @XVilka ?

Reply:
Right. Also `idx` pointer should be marked as `RZ_NONNULL` and have an assert.